aptos-move/framework/aptos-stdlib/sources/pool_u64.move

```diff
@@ -0,0 +1,535 @@
+        vector::length(&pool.shareholders)
+    }
+
+    /// Update `pool`'s total balance of coins.
+    public fun update_total_coins(pool: &mut Pool, new_total_coins: u64) {
+        pool.total_coins = new_total_coins;
+    }
+
+    /// Allow an existing or new shareholder to add their coins to the pool in exchange for new shares.
+    public fun buy_in(pool: &mut Pool, shareholder: address, coins_amount: u64): u64 {
+        if (coins_amount == 0) return 0;
+
+        let new_shares = amount_to_shares(pool, coins_amount);
+        assert!(MAX_U64 - pool.total_coins >= coins_amount, error::invalid_argument(EPOOL_TOTAL_COINS_OVERFLOW));
+        assert!(MAX_U64 - pool.total_shares >= new_shares, error::invalid_argument(EPOOL_TOTAL_COINS_OVERFLOW));
+
+        pool.total_coins = pool.total_coins + coins_amount;
+        pool.total_shares = pool.total_shares + new_shares;
+        add_shares(pool, shareholder, new_shares);
+        new_shares
+    }
+
+    /// Add the number of shares directly for `shareholder` in `pool`.
+    /// This would dilute other shareholders if the pool's balance of coins didn't change.
+    public fun add_shares(pool: &mut Pool, shareholder: address, new_shares: u64): u64 {
+        if (contains(pool, shareholder)) {
+            let existing_shares = simple_map::borrow_mut(&mut pool.shares, &shareholder);
+            let current_shares = *existing_shares;
+            assert!(MAX_U64 - current_shares >= new_shares, error::invalid_argument(ESHAREHOLDER_SHARES_OVERFLOW));
+
+            *existing_shares = current_shares + new_shares;
+            *existing_shares
+        } else if (new_shares > 0) {
+            assert!(
+                vector::length(&pool.shareholders) < pool.shareholders_limit,
+                error::invalid_state(ETOO_MANY_SHAREHOLDERS),
+            );
+
+            vector::push_back(&mut pool.shareholders, shareholder);
+            simple_map::add(&mut pool.shares, shareholder, new_shares);
+            new_shares
+        } else {
+            new_shares
+        }
+    }
+
+    /// Allow `shareholder` to redeem their shares in `pool` for coins.
+    public fun redeem_shares(pool: &mut Pool, shareholder: address, shares_to_redeem: u64): u64 {
+        assert!(contains(pool, shareholder), error::invalid_argument(ESHAREHOLDER_NOT_FOUND));
+        assert!(shares(pool, shareholder) >= shares_to_redeem, error::invalid_argument(EINSUFFICIENT_SHARES));
+
+        if (shares_to_redeem == 0) return 0;
+
+        let redeemed_coins = shares_to_amount(pool, shares_to_redeem);
+        pool.total_coins = pool.total_coins - redeemed_coins;
+        pool.total_shares = pool.total_shares - shares_to_redeem;
+        deduct_shares(pool, shareholder, shares_to_redeem);
+
+        redeemed_coins
+    }
+
+    /// Directly deduct `shareholder`'s number of shares in `pool` and return the number of remaining shares.
+    public fun deduct_shares(pool: &mut Pool, shareholder: address, num_shares: u64): u64 {
```

Comment:
I didn't see this being used externally before... so not sure why public

Reply:
Made private.